roachtest: add allocation balancing benchmark

kvoli opened the pull request:
This patch adds benchmarks to measure the performance of allocation load
balancing.

The resources that is measured to determine the balance of
load are:

- disk write bandwidth utilization [0,100]
- cpu utilization [0,100]

The balance of a cluster is equivalent to a measure of dispersion at an
instant. The measure of dispersion use is the range `(max-min)`. The
range is used as the values are normalized (utilization) in the range
`0-100` and the maximum value is significant.

The summary of a run is calculated as the arithmetic mean of balance for
each resource over the sample period. The cumulative bandwidth of
rebalancing snapshots is also included for reference.

Each benchmark samples 5 runs and selects the run which has the minimum
pairwise norm distance to other runs. The selected sample is exported in
addition to the stddev of samples. Including the stddev gives indication
of how stable the performance is.

```
write(%): avg max-min write disk utilization
cpu(%):   avg max-min cpu utilization
cost(gb): cumulative bandwidth of rebalancing snapshots

std_write(%): stddev(samples.write(%))
std_cpu(%):   stddev(samples.cpu(%))
std_cost(gb): stddev(samples.cost(gb))
```

For all these performance indicators, a lower value indicates better
performance whilst a higher number is less desirable. However, in
aggregate an appropriate goal is to reduce one indicator whilst holding
the other indicators no worse off.

resolves: https://github.com/cockroachdb/cockroach/issues/80383

Release note: None

kvoli committed:
roachtest: add allocation balancing benchmark

This patch adds benchmarks to measure the performance of allocation load
balancing.

The resources that is measured to determine the balance of
load are:

- disk write bandwidth utilization [0,100]
- cpu utilization [0,100]

The balance of a cluster is equivalent to a measure of dispersion at an
instant. The measure of dispersion use is the range `(max-min)`. The
range is used as the values are normalized (utilization) in the range
`0-100` and the maximum value is significant.

The summary of a run is calculated as the arithmetic mean of balance for
each resource over the sample period. The cumulative bandwidth of
rebalancing snapshots is also included for reference.

Each benchmark samples 5 runs and selects the run which has the minimum
pairwise norm distance to other runs. The selected sample is exported in
addition to the stddev of samples. Including the stddev gives indication
of how stable the performance is.

```
write(%): avg max-min write disk utilization
cpu(%):   avg max-min cpu utilization
cost(gb): cumulative bandwidth of rebalancing snapshots

std_write(%): stddev(samples.write(%))
std_cpu(%):   stddev(samples.cpu(%))
std_cost(gb): stddev(samples.cost(gb))
```

For all these performance indicators, a lower value indicates better
performance whilst a higher number is less desirable. However, in
aggregate an appropriate goal is to reduce one indicator whilst holding
the other indicators no worse off.

resolves: #80383

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86661)
<!-- Reviewable:end -->

kvoli commented:
Cc @tbg

kvoli commented:
These numbers are hard to tune here for workloads. It may be worth considering auto tuning to get a desired cumulative resource use that's below capacity for cpu/writes.

cc @irfansharif @andrewbaptist I used rated admission wait duration > 40s or io_overload > 1 to determine "overload". Is there something better for the non-write case? I tried using used slots/total slots but that didn't correlate as nicely as I had hoped. 

Could there be a metric ac_overload, similar to io_overload taken from AC?

lidorcarmel reviewed (dismissed):
> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND5OPgv1Sf5Ae50LIuI):*
remove the 'the'.



_Code quote:_
```
 the
```
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiTFvC5Qf-KUEJsctM:-NDiTFvC5Qf-KUEJsctN:bm4ilea) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L99)):*
> ```Go
> }
> 
> // kvAllocBenchEventRunner implements the clampResourceEventRunner interface.
> ```

wrong comment? (will fail lint I guess)
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiUxge1OPoSqxh5EDx:-NDiUxge1OPoSqxh5EDy:b-umhx14) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L161)):*
> ```Go
> 	defer db.Close()
> 
> 	// If the replication factor has been altered (!=0) or pinned attrbutes
> ```

typo attributes
and looks like we're always doing this, not only if replFactor was changed?



_Code quote:_
```Go
attrbutes
```
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiVqBB1UCBOL-SFmY3:-NDiVqBB1UCBOL-SFmY4:b-rhl1qd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L168)):*
> ```Go
> 		stmt = fmt.Sprintf("%s, constraints='[+%s]'", stmt, kvr.pinnedAttribute)
> 	}
> 	stmt = fmt.Sprintf("%s;", stmt)
> ```

nit: is `stmt += ";"` good enough?
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiXKOC21gTcYyl0TRt:-NDiXKOC21gTcYyl0TRu:b-siesz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L181)):*
> ```Go
> 
> 	// When span is specifed, add in 10% spans to the workload with 10k
> 	// spanning limit batched in sizes of 10. This puts an read (cpu) intensive
> ```

s/an/a/



_Code quote:_
```Go
an
```
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiYdCN8TknGsvEzPE1:-NDiYdCN8TknGsvEzPE2:bw7f6lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L217)):*
> ```Go
> 						readPercent: 85,
> 						rate:        650,
> 						blockSize:   1 << 12,
> ```

optional nit: I like to see `4 << 10` instead of this.. but maybe it's just me (e.g. `1 << 17` does require more thinking, right? vs simple english such as `128 << 10`)
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiZJSREDXpBmnksa1t:-NDiZJSREDXpBmnksa1u:b-7qy1oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L225)):*
> ```Go
> 		},
> 		{
> 			// NB: Skewed write operation workload, 20k ops/s vs 200 ops/s,
> ```

question: do you think we should have more "simple" values for all tests? for example, have all the read and write variations use the exact same rate and block size, so that it's easier to know what we're running, and say we only use block sizes of 1, 4kb and 256k, that's it. the way things are now we really need to read the spec every time we look at a test. but maybe you fine tuned these values? not sure why that would be needed.. anyway, we can discuss offline.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND9EcPhBFRKK_fjQZFh:-ND9EcPhBFRKK_fjQZFi:b-h105yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L543)):*
> ```Go
> 	}
> 
> 	if err := statCollector.Exporter().Export(
> ```

what do you think about exporting the cost? for example report the bytes rebalanced during the test, this is something that we should track when making improvements, so that we don't just do 10x more rebalances to make imbalance look good, and hurt pebble in the process.

I know I'm adding a metric instead of removing one as we talked :(

Regarding the other 3 metrics - unfortunately I think we need all 3 metrics that you added. the most important one is the AC one, but if we merge a pr that changes the imbalance significantly we need to know about it, even if latency and AC stays the same. I guess we should get used to exporting multiple metrics for roachperfs..
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND999Lb6OwfkVCjHX3d:-ND999Lb6OwfkVCjHX3e:bdq2sm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/87bd30727ebdea19e1d603c8ffa76d51239a84e0/pkg/cmd/roachtest/tests/rebalance_stats.go#L30)):*
> ```Go
> 	// NB: These are recorded as counters. These are then rated, as we are
> 	//     interested in the progression at points in time.
> 	ioReadStat  = clusterstats.ClusterStat{LabelName: "instance", Query: "rate(sys_host_disk_read_bytes[5m])"}
> ```

why did we lose read stat?
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND9BLmb94w1acxjEX82:-ND9BLmb94w1acxjEX83:b-zfh97l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/rebalance_stats.go#L52)):*
> ```Go
> 		// NV: We take the union of the two queries and return whichever is
> 		// greater, or both if equal (true).
> 		Query: "((storage_l0_sublevels / 20 > storage_l0_num_files / 1000) or (storage_l0_sublevels / 20 <= storage_l0_num_files / 1000)) * 100",
> ```

what is the 100 for? and this query looks like `a or !a` which is always true, what am I missing?



_Code quote:_
```Go
		// NV: We take the union of the two queries and return whichever is
		// greater, or both if equal (true).
		Query: "((storage_l0_sublevels / 20 > storage_l0_num_files / 1000) or (storage_l0_sublevels / 20 <= storage_l0_num_files / 1000)) * 100",
```


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND5OPgv1Sf5Ae50LIuI):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

remove the 'the'.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiTFvC5Qf-KUEJsctM:-NE7piZhDBN2os5XCnyF:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L99)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

wrong comment? (will fail lint I guess)
</blockquote></details>

fixed.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiUxge1OPoSqxh5EDx:-NE7q5XkDSNagUREumeZ:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L161)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

typo attributes
and looks like we're always doing this, not only if replFactor was changed?
</blockquote></details>

Updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiVqBB1UCBOL-SFmY3:-NE7qVcW5S8KQLSGbNEt:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L168)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: is `stmt += ";"` good enough?
</blockquote></details>

updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiXKOC21gTcYyl0TRt:-NE7qaPv50kUhKyXaqNL:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L181)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/an/a/
</blockquote></details>

Updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiYdCN8TknGsvEzPE1:-NE7rOL88qGZF_UHQoIq:bvnrpcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L217)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

optional nit: I like to see `4 << 10` instead of this.. but maybe it's just me (e.g. `1 << 17` does require more thinking, right? vs simple english such as `128 << 10`)
</blockquote></details>

updated
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NDiZJSREDXpBmnksa1t:-NE8QDNf5_ml1e2s748k:bv6muhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L225)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

question: do you think we should have more "simple" values for all tests? for example, have all the read and write variations use the exact same rate and block size, so that it's easier to know what we're running, and say we only use block sizes of 1, 4kb and 256k, that's it. the way things are now we really need to read the spec every time we look at a test. but maybe you fine tuned these values? not sure why that would be needed.. anyway, we can discuss offline.
</blockquote></details>

They were fine tuned, I've updated to use just four:

small reads
big reads
small writes
big writes
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND9EcPhBFRKK_fjQZFh:-NE81E5f-QgIz8B6I8G6:b-xdxsmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/allocation_bench.go#L543)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

what do you think about exporting the cost? for example report the bytes rebalanced during the test, this is something that we should track when making improvements, so that we don't just do 10x more rebalances to make imbalance look good, and hurt pebble in the process.

I know I'm adding a metric instead of removing one as we talked :(

Regarding the other 3 metrics - unfortunately I think we need all 3 metrics that you added. the most important one is the AC one, but if we merge a pr that changes the imbalance significantly we need to know about it, even if latency and AC stays the same. I guess we should get used to exporting multiple metrics for roachperfs..
</blockquote></details>

Added a cost one.
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND999Lb6OwfkVCjHX3d:-NE7rRXHB2vZpjwATr2f:bp23c8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/87bd30727ebdea19e1d603c8ffa76d51239a84e0/pkg/cmd/roachtest/tests/rebalance_stats.go#L30)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

why did we lose read stat?
</blockquote></details>

Not very meaningful in any of the tests atm as they are going to cache mainly.
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND9BLmb94w1acxjEX82:-NE7reFk6x5p11hxj1ED:bkd06ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b678d145a72282198305401fe24357628d027/pkg/cmd/roachtest/tests/rebalance_stats.go#L52)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

what is the 100 for? and this query looks like `a or !a` which is always true, what am I missing?
</blockquote></details>

Whichever is true gets returned. The patch I was waiting for merged so simplified.

100 is for scaling to be similar to other signals captured.


<!-- Sent from Reviewable.io -->

kvoli commented:
Ran the benchmarks 5x each. There is quite a bit of variability in the `rw=0` benchmarks. It may be worthwhile running each benchmark multiple times and taking a median for each value.

Results below.

<details>
    <summary>
        rw=0/access=skew
    </summary>

        "cost(mb)":939,"cpu(s) imbalance":320,"overload(s)":50,"write(s) imbalance":760
        "cost(mb)":421,"cpu(s) imbalance":10,"overload(s)":50,"write(s) imbalance":540
        "cost(mb)":1059,"cpu(s) imbalance":360,"overload(s)":50,"write(s) imbalance":760
        "cost(mb)":986,"cpu(s) imbalance":590,"overload(s)":0,"write(s) imbalance":760
        "cost(mb)":1055,"cpu(s) imbalance":680,"overload(s)":70,"write(s) imbalance":760

</details>

<details>
    <summary>
        rw=0/clamp=write
    </summary>

        "cost(mb)":125,"cpu(s) imbalance":870,"overload(s)":360,"write(s) imbalance":600
        "cost(mb)":181,"cpu(s) imbalance":140,"overload(s)":250,"write(s) imbalance":10
        "cost(mb)":94,"cpu(s) imbalance":660,"overload(s)":60,"write(s) imbalance":620
        "cost(mb)":121,"cpu(s) imbalance":880,"overload(s)":160,"write(s) imbalance":630
        "cost(mb)":82,"cpu(s) imbalance":280,"overload(s)":250,"write(s) imbalance":0

</details>

<details>
    <summary>
        rw=0/ops=skew
    </summary>
    
        "cost(mb)":460,"cpu(s) imbalance":850,"overload(s)":100,"write(s) imbalance":750
        "cost(mb)":521,"cpu(s) imbalance":800,"overload(s)":50,"write(s) imbalance":720
        "cost(mb)":702,"cpu(s) imbalance":640,"overload(s)":60,"write(s) imbalance":750
        "cost(mb)":505,"cpu(s) imbalance":690,"overload(s)":90,"write(s) imbalance":750
        "cost(mb)":703,"cpu(s) imbalance":530,"overload(s)":80,"write(s) imbalance":660

</details>

<details>
    <summary>
        rw=0/ops=skew/pin
    </summary>

        "cost(mb)":343,"cpu(s) imbalance":640,"overload(s)":110,"write(s) imbalance":720
        "cost(mb)":461,"cpu(s) imbalance":850,"overload(s)":50,"write(s) imbalance":710
        "cost(mb)":329,"cpu(s) imbalance":800,"overload(s)":80,"write(s) imbalance":720
        "cost(mb)":343,"cpu(s) imbalance":870,"overload(s)":100,"write(s) imbalance":720
        "cost(mb)":374,"cpu(s) imbalance":870,"overload(s)":60,"write(s) imbalance":720

</details>

<details>
    <summary>
        rw=95/access=skew
    </summary>

        "cost(mb)":0.2,"cpu(s) imbalance":200,"overload(s)":800,"write(s) imbalance":0
        "cost(mb)":0.3,"cpu(s) imbalance":220,"overload(s)":570,"write(s) imbalance":0
        "cost(mb)":0.3,"cpu(s) imbalance":310,"overload(s)":600,"write(s) imbalance":0
        "cost(mb)":0.2,"cpu(s) imbalance":190,"overload(s)":390,"write(s) imbalance":0
        "cost(mb)":0.7,"cpu(s) imbalance":120,"overload(s)":320,"write(s) imbalance":0

</details>

<details>
    <summary>
        rw=95/clamp=cpu
    </summary>

        "cost(mb)":0.2,"cpu(s) imbalance":130,"overload(s)":870,"write(s) imbalance":0
        "cost(mb)":0.5,"cpu(s) imbalance":130,"overload(s)":870,"write(s) imbalance":0
        "cost(mb)":0.6,"cpu(s) imbalance":210,"overload(s)":870,"write(s) imbalance":0
        "cost(mb)":0.6,"cpu(s) imbalance":130,"overload(s)":860,"write(s) imbalance":0
        "cost(mb)":0.2,"cpu(s) imbalance":130,"overload(s)":870,"write(s) imbalance":0

</details>

<details>
    <summary>
        rw=95/ops=skew
    </summary>

        "cost(mb)":0.3,"cpu(s) imbalance":880,"overload(s)":650,"write(s) imbalance":0
        "cost(mb)":0.2,"cpu(s) imbalance":870,"overload(s)":710,"write(s) imbalance":0
        "cost(mb)":0.4,"cpu(s) imbalance":870,"overload(s)":820,"write(s) imbalance":0
        "cost(mb)":0.1,"cpu(s) imbalance":880,"overload(s)":530,"write(s) imbalance":0
        "cost(mb)":0.4,"cpu(s) imbalance":870,"overload(s)":710,"write(s) imbalance":0

</details>

<details>
    <summary>
        rw=95/ops=skew/pin
    </summary>

        "cost(mb)":0.6,"cpu(s) imbalance":880,"overload(s)":640,"write(s) imbalance":0
        "cost(mb)":0.6,"cpu(s) imbalance":880,"overload(s)":470,"write(s) imbalance":0
        "cost(mb)":0.5,"cpu(s) imbalance":880,"overload(s)":610,"write(s) imbalance":0
        "cost(mb)":0.4,"cpu(s) imbalance":870,"overload(s)":510,"write(s) imbalance":0
        "cost(mb)":0.2,"cpu(s) imbalance":880,"overload(s)":700,"write(s) imbalance":0

</details>

kvoli commented:
Updated this to take the "middle" run of multiple runs. Seems more stable than before.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 4 files at r5, 2 of 5 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-:-NKZ5uroD7L6QZyvLZe4:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @kvoli, and @lidorcarmel)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND5OPgv1Sf5Ae50LIuI):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

done
</blockquote></details>

Was this done?
___
*[`-- commits` line 4 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKYyiZE9aLQH-3oOwBY):*
"benchmarking"
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 82 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ6P-wCCV6MR-hOWHf:-NKZ6P-wCCV6MR-hOWHg:b-p367wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L82)):*
> ```Go
> type allocBenchLoadEvent struct {
> 	run       allocBenchEventRunner
> 	startTime time.Duration
> ```

It doesn't look like any of the tests are currently playing with temporal skew across workloads, either by starting one workload before the other or ending one workload before the other. That might be interesting as we explore the reactivity of load statistics to changes in workloads. Do you plan to add variants that do so in the future?
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 195 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ1BwD3rBMH69vj_Uu:-NKZ1BwD3rBMH69vj_Uv:b-pjghsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L195)):*
> ```Go
> 		stmt = fmt.Sprintf("%s, constraints='[+%s]'", stmt, r.pinnedAttribute)
> 	}
> 	stmt += ";"
> ```

nit: you shouldn't need to terminate the statement with a semicolon.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ12Vf3nkFjux0eokv:-NKZ12Vf3nkFjux0eokw:b-uim3x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L207)):*
> ```Go
> 	}
> 
> 	// When span is specifed, add in 10% spans to the workload with 10k
> ```

`20%`?
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 226 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ1mhU9qouRbql1ewX:-NKZ1mhU9qouRbql1ewY:b-gtnx0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L226)):*
> ```Go
> 		// import and index backfills.
> 		{
> 			// NB: Skewed read operation workload, 20k ops/s vs 650 ops/s,
> ```

nit: these ops/s values don't line up with the definitions above.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 294 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ2BBB3h--Iknyp8n6:-NKZ2BBB3h--Iknyp8n7:buvuxfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L294)):*
> The remaining load is not pinned

As a result, we expect the load to shift onto the other 4 nodes, where there is sufficient capacity. Is that right? The next sentence ("with the pinned load being immovable") has me thinking there's a constraint that makes this not all balance out well, but you then say the cluster is overprovisioned, not underprovisioned.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 429 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ4pGeDeVmRlReffp2:-NKZ4pGeDeVmRlReffp3:bo9cypd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L429)):*
> ```Go
> 		// results of future runs.
> 		cleanupFunc()
> 		c.Wipe(ctx)
> ```

nit: any reason not to push `c.Wipe` into `cleanupFunc`?
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 433 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ51afBv2G0BV5-ItT:-NKZ51afBv2G0BV5-ItU:bg5afym) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L433)):*
> ```Go
> 	// Find the middle run to export. The test has high variation due to
> 	// generally compounding effects of bad decisions earlier on that are
> 	// random.
> ```

Do we find that grabbing the median run is sufficient to avoid the variability across runs of this test?
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 460 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ3MCF4OtCOa4lxuAQ:-NKZ3MCF4OtCOa4lxuAR:brw1a3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L460)):*
> ```Go
> 	specLoad := &spec.load
> 	m := c.NewMonitor(ctx, c.Nodes(1, spec.nodes))
> 	for range spec.load.events {
> ```

In my opinion, this allocation of events to goroutines through the atomic access on `loadConf.eventCounter` is more complex than it needs to be. It's also more error-prone than it needs to be, requiring mutability on the specification and `len(spec.load.events)` or fewer calls to `runAllocationBenchEvent` to avoid a panic.

Can we simplify this by using the loop iteration index of the loop that we're already running?
```go
for i := range spec.load.events {
    i := i // copy loop variable
    m.Go(func(ctx context.Context) error {
        return runAllocationBenchEvent(ctx, t, c, spec.load, i)
    })
}
```
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 601 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ59DU3hfe03gzayV5:-NKZ59DU3hfe03gzayV6:b-uofsf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L601)):*
> ```Go
> }
> 
> // conbinePenalties collects the penalty arrays for each threshold given.
> ```

`combinePenalties`
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 91 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ-ZVV1wb0pOcruWd6:-NKZ-ZVV1wb0pOcruWd7:b6ux8w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/rebalance_stats.go#L91)):*
> ```Go
> }
> 
> func divQuery(query string, max int) string {
> ```

nit: is `max` the right name for this param? `val`?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86661)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @lidorcarmel, and @nvanbenschoten)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-ND5OPgv1Sf5Ae50LIuI):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this done?
</blockquote></details>

I think I reverted to an older commit message somewhere. Fixed now.
___
*[`-- commits` line 4 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKYyiZE9aLQH-3oOwBY):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"benchmarking"
</blockquote></details>

fixed.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 82 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ6P-wCCV6MR-hOWHf:-NKslNMu1b7bqVlbtgx0:b3fri12) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L82)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It doesn't look like any of the tests are currently playing with temporal skew across workloads, either by starting one workload before the other or ending one workload before the other. That might be interesting as we explore the reactivity of load statistics to changes in workloads. Do you plan to add variants that do so in the future?
</blockquote></details>

That is a good idea. I hadn't planned on including that but will add an issue for this.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 195 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ1BwD3rBMH69vj_Uu:-NKslgtm0cqZ5FijJ3QE:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L195)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: you shouldn't need to terminate the statement with a semicolon.
</blockquote></details>

updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ12Vf3nkFjux0eokv:-NKslkRK0y3Hn5mYyw3e:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L207)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`20%`?
</blockquote></details>

Updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 226 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ1mhU9qouRbql1ewX:-NKsm9v9DuwVOq7V4nCK:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L226)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: these ops/s values don't line up with the definitions above.
</blockquote></details>

fixed.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 294 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ2BBB3h--Iknyp8n6:-NKsmH-P4MjmWOt_WC05:b-uj1w98) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L294)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> The remaining load is not pinned

As a result, we expect the load to shift onto the other 4 nodes, where there is sufficient capacity. Is that right? The next sentence ("with the pinned load being immovable") has me thinking there's a constraint that makes this not all balance out well, but you then say the cluster is overprovisioned, not underprovisioned.
</blockquote></details>

The cluster has enough total resources to manage if it balances correctly with pinned vs unpinned load. Updated this to be clearer.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 429 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ4pGeDeVmRlReffp2:-NKsmwSVC2kiUPRBiEwg:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L429)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: any reason not to push `c.Wipe` into `cleanupFunc`?
</blockquote></details>

Updated.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 433 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ51afBv2G0BV5-ItT:-NKsogMoCGwK0IfHcdrm:b-i0k3u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L433)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we find that grabbing the median run is sufficient to avoid the variability across runs of this test?
</blockquote></details>

It did when I last had run these. I'm rerunning now 4x to confirm.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 460 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ3MCF4OtCOa4lxuAQ:-NKsq7JsAmMJGakfZPd1:b-p3uew3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L460)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In my opinion, this allocation of events to goroutines through the atomic access on `loadConf.eventCounter` is more complex than it needs to be. It's also more error-prone than it needs to be, requiring mutability on the specification and `len(spec.load.events)` or fewer calls to `runAllocationBenchEvent` to avoid a panic.

Can we simplify this by using the loop iteration index of the loop that we're already running?
```go
for i := range spec.load.events {
    i := i // copy loop variable
    m.Go(func(ctx context.Context) error {
        return runAllocationBenchEvent(ctx, t, c, spec.load, i)
    })
}
```
</blockquote></details>

Simplified - good suggestion.
___
*[`pkg/cmd/roachtest/tests/allocation_bench.go` line 601 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ59DU3hfe03gzayV5:-NKsn3NFATFUfdweqLAE:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/allocation_bench.go#L601)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`combinePenalties`
</blockquote></details>

Updated.
___
*[`pkg/cmd/roachtest/tests/rebalance_stats.go` line 91 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-NKZ-ZVV1wb0pOcruWd6:-NKsnEHT4vQu2lECSGa4:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e45188609dbb8a545e34967ad838b998a3d9a5/pkg/cmd/roachtest/tests/rebalance_stats.go#L91)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: is `max` the right name for this param? `val`?
</blockquote></details>

Updated.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86661#-:-NKtKxv671z1cyaNnRjM:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @lidorcarmel)



<!-- Sent from Reviewable.io -->

kvoli commented:
I've updated to use a more stable summary metric, the average range of CPU/Write bandwidth over a run. I've also updated to include the stddev in the output results.

The previous penalty approach required defining a threshold for penalty and in cases where the cluster would hover just above the threshold it would lead to a high variance in results. This change helps produce less variable results.

bors r+

kvoli commented:
bors r+
